# **Author: Alejandro C. Parra Garcia**

In [11]:
import matplotlib.pyplot as plt
import numpy as np
import math
%matplotlib inline

# Functions

## Newton method

In [2]:
def Newton_Method(F, DF, x, tol=1e-5):
    norm = np.linalg.norm(F(x))
    print(f'x = {x}, {norm = :.6f}')
    while norm > tol:
        s = np.linalg.inv(DF(x)).dot(F(x))        
        x -= s
        norm = np.linalg.norm(F(x))
        print(f'x = {x}, {norm = :.6f}')
    
    return x

## Gradient Descent

In [5]:
# Program 13.1 Golden Section Search for minimum of f(x)
# Start with unimodal f(x) and minimum in [a,b]
# Input: function f, interval [a,b], desired accuracy tol
# Output: approximate minimum y
def GSS(f, a, b, tol=1e-5):
    g=(np.sqrt(5)-1)/2   # golden mean
    x1 = a+(1-g)*(b-a)
    x2 = a+g*(b-a)
    f1=f(x1)
    f2=f(x2)
    while (b-a)>tol:
        if f1 < f2: # if f(x1) < f(x2), replace b with x2
            b=x2
            x2=x1
            x1=a+(1-g)*(b-a)
            f2=f1
            f1=f(x1) # single function evaluation
        else: # otherwise, replace a with x1
            a=x1
            x1=x2
            x2=a+g*(b-a)
            f1=f2
            f2=f(x2) # single function evaluation
        #print("(a, b): (%.10f, %.10f)" % (a, b))
    y=(a+b)/2;
    return y

# Gradient (Steepest) Descent method for minimum of f(x)
# Input: 
#    function f:R^n \to R, 
#    gradient grad:R^n \to R^n, 
#    initial guess x, 
#    desired accuracy tol
# Output: approximate minimum x \in R^n
def GradientDescent(f, grad, x, tol=1e-5):
    # IMPLEMENT GRADIENT DESCENT HERE!
    
    while abs(np.linalg.norm(grad(x)))>tol:
        
        def F_n(s):
            val = x-s*grad(x)
            return f(val[0],val[1])
        
        s = GSS(F_n, a=0, b=1)
        
        x = x - s * grad(x)
        
        print("(x, y): (%.10f, %.10f)" % (x[0], x[1]))
    return x

# PROBLEMS

## Computer Problems 1

Use Newton’s Method to find the minimum of:
$$ f (x, y) = e^{−x^2 y^2} + (x − 1)^2 + (y − 1)^2 $$

Try various initial conditions, and compare answers. How many correct digits can you
obtain with this method?

### Solution.

**derivatives:**

partial of f with respect to x:
$$ -2y^2xe^{−x^2 y^2} + 2(x − 1)  $$ 

partial of f with respect to y:
$$ -2x^2ye^{−x^2 y^2} + 2(y − 1)  $$ 

**Second derivatives:**

partial of f with respect to x x:
$$ 2y^2e^{−x^2 y^2}(2y^2x^2-1)+2 $$

partial of f with respect to x y:
$$ 4xye^{−x^2 y^2}(x^2y^2-1) $$

partial of f with respect to y x:
$$ 4xye^{−x^2 y^2}(x^2y^2-1) $$

partial of f with respect to y y:
$$ 2x^2e^{−x^2 y^2}(2x^2y^2-1)+2 $$


In [16]:
# We are looking for critical point of the gradient, i.e.
# F = grad(f)
# DF = Hess(f)

def F_Prob1(X):
    x = X[0]
    y = X[1]
        
    f1 = 2*(x-1) - 2 * y**2 * x * math.exp(-x**2 * y**2)   # partial of f with respect to x
    f2 = 2*(y-1) - 2 * x**2 * y * math.exp(-x**2 * y**2)   # partial of f with respect to y
    
    return np.array([f1, f2])
        
def DF_Prob1(X):
    x = X[0]
    y = X[1]
    
    df11 = 2 * y**2 * math.exp(-x**2 * y**2) * (2 * y**2 * x**2 - 1) +2     # partial of f with respect to x x
    df12 = 4 * x * y * math.exp(-x**2 * y**2) * (x**2 * y**2 -1)            # partial of f with respect to x y
    df21 = 4 * x * y * math.exp(-x**2 * y**2) * (x**2 * y**2 -1)            # partial of f with respect to y x
    df22 = 2 * x**2 * math.exp(-x**2 * y**2) * (2 * x**2 * y**2 - 1) +2     # partial of f with respect to y y
    
    df = np.array([[df11, df12], [df21, df22]])
    
    return df

In [22]:
#initial_guess = np.array([1.0, 1.0])
initial_guesses = [np.array([1.0, 1.0]), np.array([-1.0, -1.0]), np.array([-5.0, 5.0]), np.array([0.0, 0.0])]
for i, initial_guess in enumerate(initial_guesses):
    print("Guess nº " + str(i) + " | Initial condition: " + str(initial_guess))
    (x, y) = Newton_Method(F_Prob1, DF_Prob1, initial_guess)
    print(" ")

Guess nº 0 | Initial condition: [1. 1.]
x = [1. 1.], norm = 1.040520
x = [1.26894142 1.26894142], norm = 0.328329
x = [1.20745566 1.20745566], norm = 0.007540
x = [1.20881746 1.20881746], norm = 0.000001
 
Guess nº 1 | Initial condition: [-1. -1.]
x = [-1. -1.], norm = 4.616334
x = [0.19317574 0.19317574], norm = 2.302405
x = [1.10944234 1.10944234], norm = 0.539423
x = [1.21377776 1.21377776], norm = 0.027446
x = [1.20881508 1.20881508], norm = 0.000014
x = [1.20881759 1.20881759], norm = 0.000000
 
Guess nº 2 | Initial condition: [-5.  5.]
x = [-5.  5.], norm = 14.422205
x = [1. 1.], norm = 1.040520
x = [1.26894142 1.26894142], norm = 0.328329
x = [1.20745566 1.20745566], norm = 0.007540
x = [1.20881746 1.20881746], norm = 0.000001
 
Guess nº 3 | Initial condition: [0. 0.]
x = [0. 0.], norm = 2.828427
x = [1. 1.], norm = 1.040520
x = [1.26894142 1.26894142], norm = 0.328329
x = [1.20745566 1.20745566], norm = 0.007540
x = [1.20881746 1.20881746], norm = 0.000001
 


The result is **[1.20881759 1.20881759]**, and its a similar result for all the initial conditions, this one is the solution with the lowest norm. and thus has a lower error.

## Computer Problems 2

Apply Newton’s Method to find the minima of the following functions to six correct
decimal places (each function has two minima):
- (a) $$ f (x, y) = x^4 + y^4 + 2x^2 y^2 + 6x y − 4x − 4y + 1 $$
- (b) $$ f (x, y) = x^6 + y^6 + 3x^2 y^2 − x^2 − y^2 − 2x y $$

### Solution.

#### Equation A:

**derivatives for eq a:**

partial of f with respect to x:
$$ 4x^3+4y^2x+6y-4  $$ 

partial of f with respect to y:
$$ 4y^3+4x^2y+6x-4  $$ 

**Second derivatives for eq a:**

partial of f with respect to x x:
$$ 12x^2+4y^2 $$

partial of f with respect to x y:
$$ 8xy+6 $$

partial of f with respect to y x:
$$ 8xy+6 $$

partial of f with respect to y y:
$$ 12y^2+4x^2 $$


In [60]:
# We are looking for critical point of the gradient, i.e.
# F = grad(f)
# DF = Hess(f)

def F_Prob2a(X):
    x = X[0]
    y = X[1]
        
    f1 = 4*x**3 + 4*y**2*x + 6*y - 4   # partial of f with respect to x
    f2 = 4*y**3 + 4*x**2*y + 6*x - 4   # partial of f with respect to y
    
    return np.array([f1, f2])
        
def DF_Prob2a(X):
    x = X[0]
    y = X[1]
    
    df11 = 12*x**2 + 4*y**2     # partial of f with respect to x x
    df12 = 8*x*y + 6            # partial of f with respect to x y
    df21 = 8*x*y + 6            # partial of f with respect to y x
    df22 = 12*y**2 + 4*x**2     # partial of f with respect to y y
    
    df = np.array([[df11, df12], [df21, df22]])
    
    return df

In [26]:
#initial_guess = np.array([1.0, 1.0])
initial_guesses = [np.array([1.0, 1.0]), np.array([-1.0, -1.0]), np.array([-5.0, 5.0]), np.array([0.0, 0.0])]
for i, initial_guess in enumerate(initial_guesses):
    print("Guess nº " + str(i) + " | Initial condition: " + str(initial_guess))
    (x, y) = Newton_Method(F_Prob2a, DF_Prob2a, initial_guess)
    print(" ")

Guess nº 0 | Initial condition: [1. 1.]
x = [1. 1.], norm = 14.142136
x = [0.66666667 0.66666667], norm = 3.352210
x = [0.52444444 0.52444444], norm = 0.425142
x = [0.50058758 0.50058758], norm = 0.009977
x = [0.50000035 0.50000035], norm = 0.000006
 
Guess nº 1 | Initial condition: [-1. -1.]
x = [-1. -1.], norm = 25.455844
x = [-0.4 -0.4], norm = 9.775044
x = [0.30243902 0.30243902], norm = 2.777593
x = [0.54209628 0.54209628], norm = 0.745315
x = [0.5017206 0.5017206], norm = 0.029250
x = [0.50000296 0.50000296], norm = 0.000050
x = [0.5 0.5], norm = 0.000000
 
Guess nº 2 | Initial condition: [-5.  5.]
x = [-5.  5.], norm = 1371.798819
x = [-3.34758589  3.38642084], norm = 403.308259
x = [-2.24216737  2.34808708], norm = 117.396766
x = [-1.49422634  1.71594651], norm = 33.420203
x = [-0.98410992  1.3733048 ], norm = 9.072274
x = [-0.65893905  1.21444262], norm = 2.188290
x = [-0.50419639  1.14971496], norm = 0.358106
x = [-0.46794496  1.1335832 ], norm = 0.018085
x = [-0.46597777  1.

In [69]:
print(F_Prob2a([0.5, 0.5]))
print(F_Prob2a([-0.46597192,  1.13263859]))

[0. 0.]
[9.21615317e-09 4.62123131e-08]


For **equation A** we can see that there are 2 solutions, at **[0.5 0.5]** and **[-0.46597192  1.13263859]**. We can see that the first solution is exact, while the second one is aproximate, since the derivative at that point returns a number close to 0, but not exactly 0.

#### Equation B:

**derivatives for eq b:**

partial of f with respect to x:
$$ 6x^5+6y^2x-2x-2y  $$ 

partial of f with respect to y:
$$ 6y^5+6x^2y-2y-2x  $$ 

**Second derivatives for eq b:**

partial of f with respect to x x:
$$ 30x^4+6y^2-2 $$

partial of f with respect to x y:
$$ 12xy-2 $$

partial of f with respect to y x:
$$ 12xy-2 $$

partial of f with respect to y y:
$$ 30y^4+6x^2-2 $$


In [30]:
# We are looking for critical point of the gradient, i.e.
# F = grad(f)
# DF = Hess(f)

def F_Prob2b(X):
    x = X[0]
    y = X[1]
        
    f1 = 6*x**5 + 6*y**2*x - 2*x -2*y   # partial of f with respect to x
    f2 = 6*y**5 + 6*x**2*y - 2*y -2*x   # partial of f with respect to y
    
    return np.array([f1, f2])
        
def DF_Prob2b(X):
    x = X[0]
    y = X[1]
    
    df11 = 30*x**4 + 6*y**2 -2     # partial of f with respect to x x
    df12 = 12*x*y - 2              # partial of f with respect to x y
    df21 = 12*x*y - 2              # partial of f with respect to y x
    df22 = 30*y**4 + 6*x**2 -2     # partial of f with respect to y y
    
    df = np.array([[df11, df12], [df21, df22]])
    
    return df

In [31]:
#initial_guess = np.array([1.0, 1.0])
initial_guesses = [np.array([1.0, 1.0]), np.array([-1.0, -1.0]), np.array([-5.0, 5.0]), np.array([0.0, 0.0])]
for i, initial_guess in enumerate(initial_guesses):
    print("Guess nº " + str(i) + " | Initial condition: " + str(initial_guess))
    (x, y) = Newton_Method(F_Prob2b, DF_Prob2b, initial_guess)
    print(" ")

Guess nº 0 | Initial condition: [1. 1.]
x = [1. 1.], norm = 11.313708
x = [0.81818182 0.81818182], norm = 3.130235
x = [0.71520059 0.71520059], norm = 0.646251
x = [0.68020139 0.68020139], norm = 0.058147
x = [0.67637671 0.67637671], norm = 0.000641
x = [0.67633358 0.67633358], norm = 0.000000
 
Guess nº 1 | Initial condition: [-1. -1.]
x = [-1. -1.], norm = 11.313708
x = [-0.81818182 -0.81818182], norm = 3.130235
x = [-0.71520059 -0.71520059], norm = 0.646251
x = [-0.68020139 -0.68020139], norm = 0.058147
x = [-0.67637671 -0.67637671], norm = 0.000641
x = [-0.67633358 -0.67633358], norm = 0.000000
 
Guess nº 2 | Initial condition: [-5.  5.]
x = [-5.  5.], norm = 27577.164466
x = [-3.984375  3.984375], norm = 9057.262030
x = [-3.16815279  3.16815279], norm = 2978.126933
x = [-2.51069525  2.51069525], norm = 980.809518
x = [-1.97946183  1.97946183], norm = 323.683030
x = [-1.54852132  1.54852132], norm = 107.060864
x = [-1.19749448  1.19749448], norm = 35.465504
x = [-0.91089633  0.9108

In [33]:
print(F_Prob2b([0.0, 0.0]))
print(F_Prob2b([-0.67633358, -0.67633358]))

[0. 0.]
[-2.8589763e-08 -2.8589763e-08]


For **equation B** we can see that there are 2 solutions, at **[0.0 0.0]** and **[-0.67633358 -0.67633358]**. And as well as before we can see that the first solution is exact, while the second one is aproximate, since the derivative at that point returns a number close to 0, but not exactly 0.

## Computer Problems 3

Find the minimum of the Rosenbrock function 
$$ f (x, y) = 100(y − x^2)^2 + (x − 1)^2 $$ 
by
+ (a) Newton’s Method. 
+ (b) Steepest Descent. 

Use starting guess (2,2). After how many steps does the solution stop improving? Explain the difference in accuracy that is achieved.

### Solution.

#### (a) Newton’s Method

**derivatives:**

partial of f with respect to x:
$$ 2(x-1)-400x(y-x^2)  $$ 

partial of f with respect to y:
$$ 200(y-x^2)  $$ 

**Second derivatives:**

partial of f with respect to x x:
$$ 1200x^2-400y+2 $$

partial of f with respect to x y:
$$ -400x $$

partial of f with respect to y x:
$$ -400x $$

partial of f with respect to y y:
$$ 200 $$


In [34]:
# We are looking for critical point of the gradient, i.e.
# F = grad(f)
# DF = Hess(f)

def F_Prob3(X):
    x = X[0]
    y = X[1]
        
    f1 = 2*(x-1) - 400*x*(y-x**2)   # partial of f with respect to x
    f2 = 200*(y-x**2)               # partial of f with respect to y
    
    return np.array([f1, f2])
        
def DF_Prob3(X):
    x = X[0]
    y = X[1]
    
    df11 = 1200*x**2 - 400*y +2     # partial of f with respect to x x
    df12 = -400*x                   # partial of f with respect to x y
    df21 = -400*x                   # partial of f with respect to y x
    df22 = 200                      # partial of f with respect to y y
    
    df = np.array([[df11, df12], [df21, df22]])
    
    return df

In [37]:
initial_guess = np.array([2.0, 2.0])
(x, y) = Newton_Method(F_Prob3, DF_Prob3, initial_guess)


x = [2. 2.], norm = 1651.182606
x = [1.99750623 3.99002494], norm = 1.999982
x = [1.00123913 0.00993165], norm = 444.323316
x = [1.00123292 1.00246736], norm = 0.002466
x = [1.         0.99999848], norm = 0.000680
x = [1. 1.], norm = 0.000000


In [43]:
print(F_Prob3([1.0, 1.0]))

[0. 0.]


#### (b) Steepest Descent

In [40]:
def f_Prob3(x,y):
    return 100*(y-x**2)**2 + (x-1)**2

def grad_Prob3(X):
    x = X[0]
    y = X[1]
    
    f1 = 2*(x-1) - 400*x*(y-x**2)   # partial of f with respect to x
    f2 = 200*(y-x**2)               # partial of f with respect to y
    
    return np.array([f1, f2])

In [41]:

initial_guess = np.array([2.0, 2.0])

GradientDescent(f_Prob3, grad_Prob3, initial_guess)

(x, y): (-1.7120885846, 2.9268635667)
(x, y): (1.8142414276, 3.2937746791)
(x, y): (1.8143618296, 3.2924728472)
(x, y): (1.8138722822, 3.2924275621)
(x, y): (1.8139842403, 3.2910507126)
(x, y): (1.8134797823, 3.2910096116)
(x, y): (1.8135991524, 3.2897000517)
(x, y): (1.8131084497, 3.2896552350)
(x, y): (1.8132194159, 3.2882706921)
(x, y): (1.8127138067, 3.2882300605)
(x, y): (1.8128322518, 3.2869127566)
(x, y): (1.8123402963, 3.2868684363)
(x, y): (1.8124503186, 3.2854762004)
(x, y): (1.8119435185, 3.2854360505)
(x, y): (1.8120611170, 3.2841099742)
(x, y): (1.8115675296, 3.2840662561)
(x, y): (1.8116768129, 3.2826679351)
(x, y): (1.8111691222, 3.2826281853)
(x, y): (1.8112856721, 3.2812917567)
(x, y): (1.8107901638, 3.2812487221)
(x, y): (1.8108988228, 3.2798459462)
(x, y): (1.8103906303, 3.2798064904)
(x, y): (1.8105056879, 3.2784598486)
(x, y): (1.8100087128, 3.2784173734)
(x, y): (1.8101164900, 3.2770069267)
(x, y): (1.8096070674, 3.2769679671)
(x, y): (1.8097211387, 3.2756126378)


array([1.00001088, 1.00002181])

In [45]:
print(grad_Prob3([1.00001088, 1.00002181]))

[1.80713272e-06 9.97632510e-06]


We can easily see that newton method is faster and more precise, in this case, than gradient descent. Newton method took 5 iteration to achive a solution while newton took a lot longer. It is also interesting to see that newton method was able to get the correct solution, while the gradient descent was not able. This is beacuse the gradient descent is not ment to have a higher accuracy than newton, but to be able to be used in higher dimensions with more complicated functions, because in newton method we need to compute the hessian, and this is quite difficult with higher dimension and complicated functions.

## Computer Problems 4

Use the Steepest Descent to find the minima of the functions in Computer Problem 2:
- (a) $$ f (x, y) = x^4 + y^4 + 2x^2 y^2 + 6x y − 4x − 4y + 1 $$
- (b) $$ f (x, y) = x^6 + y^6 + 3x^2 y^2 − x^2 − y^2 − 2x y $$

#### Equation A:

In [46]:
def f_Prob4a(x,y):
    return x**4 + y**4 + 2*x**2*y**2 + 6*x*y - 4*x - 4*y +1

def grad_Prob4a(X):
    x = X[0]
    y = X[1]
    
    f1 = 4*x**3 + 4*y**2*x + 6*y - 4   # partial of f with respect to x
    f2 = 4*y**3 + 4*x**2*y + 6*x - 4   # partial of f with respect to y
    
    return np.array([f1, f2])

In [47]:

initial_guesses = [np.array([1.0, 1.0]), np.array([-1.0, -1.0]), np.array([-5.0, 5.0]), np.array([0.0, 0.0])]
for i, initial_guess in enumerate(initial_guesses):
    print("Guess nº " + str(i) + " | Initial condition: " + str(initial_guess))
    GradientDescent(f_Prob4a, grad_Prob4a, initial_guess)
    print(" ")

Guess nº 0 | Initial condition: [1. 1.]
(x, y): (0.4999997313, 0.4999997313)
 
Guess nº 1 | Initial condition: [-1. -1.]
(x, y): (0.4999650411, 0.4999650411)
(x, y): (0.5000000030, 0.5000000030)
 
Guess nº 2 | Initial condition: [-5.  5.]
(x, y): (0.8916802779, -0.8432886534)
(x, y): (1.1790466695, -0.5525531543)
(x, y): (1.1205656396, -0.4947382675)
(x, y): (1.1383652545, -0.4767336563)
(x, y): (1.1311594689, -0.4696090661)
(x, y): (1.1333789085, -0.4673642299)
(x, y): (1.1324476037, -0.4664433454)
(x, y): (1.1327348039, -0.4661528984)
(x, y): (1.1326137636, -0.4660332044)
(x, y): (1.1326511040, -0.4659954517)
(x, y): (1.1326353545, -0.4659798889)
(x, y): (1.1326402097, -0.4659749857)
(x, y): (1.1326381671, -0.4659729535)
(x, y): (1.1326387994, -0.4659723161)
 
Guess nº 3 | Initial condition: [0. 0.]
(x, y): (0.4999906239, 0.4999906239)
(x, y): (0.5000000039, 0.5000000039)
 


In [53]:
print(grad_Prob4a([0.5000000030, 0.5000000030]))
print(grad_Prob4a([1.1326387994, -0.4659723161]))

[3.60000003e-08 3.60000003e-08]
[ 2.74750028e-06 -2.68315991e-06]


We can see that the minima for **equation A** are **[0.5000000030, 0.5000000030]** and **[1.1326387994, -0.4659723161]**

#### Equation B:

In [48]:
def f_Prob4b(x,y):
    return x**6 + y**6 + 3*x**2*y**2 - x**2 - y**2 - 2*x*y

def grad_Prob4b(X):
    x = X[0]
    y = X[1]
    
    f1 = 6*x**5 + 6*y**2*x - 2*x -2*y   # partial of f with respect to x
    f2 = 6*y**5 + 6*x**2*y - 2*y -2*x   # partial of f with respect to y
    
    return np.array([f1, f2])

In [49]:
initial_guesses = [np.array([1.0, 1.0]), np.array([-1.0, -1.0]), np.array([-5.0, 5.0]), np.array([0.0, 0.0])]
for i, initial_guess in enumerate(initial_guesses):
    print("Guess nº " + str(i) + " | Initial condition: " + str(initial_guess))
    GradientDescent(f_Prob4b, grad_Prob4b, initial_guess)
    print(" ")

Guess nº 0 | Initial condition: [1. 1.]
(x, y): (-0.6763516345, -0.6763516345)
(x, y): (-0.6763335745, -0.6763335745)
 
Guess nº 1 | Initial condition: [-1. -1.]
(x, y): (0.6763516345, 0.6763516345)
(x, y): (0.6763335745, 0.6763335745)
 
Guess nº 2 | Initial condition: [-5.  5.]
(x, y): (-0.0539313417, 0.0539313417)
(x, y): (-0.0529874238, 0.0529874238)
(x, y): (-0.0520922957, 0.0520922957)
(x, y): (-0.0512418501, 0.0512418501)
(x, y): (-0.0504324515, 0.0504324515)
(x, y): (-0.0496608686, 0.0496608686)
(x, y): (-0.0489242176, 0.0489242176)
(x, y): (-0.0482199153, 0.0482199153)
(x, y): (-0.0475456402, 0.0475456402)
(x, y): (-0.0468992989, 0.0468992989)
(x, y): (-0.0462789980, 0.0462789980)
(x, y): (-0.0456830201, 0.0456830201)
(x, y): (-0.0451098032, 0.0451098032)
(x, y): (-0.0445579230, 0.0445579230)
(x, y): (-0.0440260777, 0.0440260777)
(x, y): (-0.0435130744, 0.0435130744)
(x, y): (-0.0430178182, 0.0430178182)
(x, y): (-0.0425393013, 0.0425393013)
(x, y): (-0.0420765950, 0.0420765950

In [56]:
print(grad_Prob4b([-0.0105594064, 0.0105594064]))
print(grad_Prob4b([-0.6763335745, -0.6763335745]))

[-7.06509794e-06  7.06509794e-06]
[2.92200482e-08 2.92200482e-08]


We can see that the minima for **equation B** are **[-0.0105594064, 0.0105594064]** and **[-0.6763335745, -0.6763335745]**